core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1038,7 +1035,11 @@ private[spark] class TaskSetManager(
     val minFinishedForSpeculation = (speculationQuantile * numTasks).floor.toInt
     logDebug("Checking for speculative tasks: minFinished = " + minFinishedForSpeculation)
 
-    if (tasksSuccessful >= minFinishedForSpeculation && tasksSuccessful > 0) {
+    // It's possible that a task is marked as completed by the scheduler, then the size of
+    // `successfulTaskDurations` may not equal to `tasksSuccessful`. Here we should only count the
+    // tasks that are submitted by this `TaskSetManager` and are completed successfully.
+    val numSuccessfulTasks = successfulTaskDurations.size()
+    if (numSuccessfulTasks >= minFinishedForSpeculation && numSuccessfulTasks > 0) {
```

Comment:
unrelated to your change, but a potential cleanup here -- why not compute `minFinishedForSpeculation` in the construction, and even set it to `math.max(1, (speculationQuantile * numTasks).floor.toInt)`?  Then you also don't need the extra `numSuccessfulTasks > 0` here.

Reply:
+1